group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/SubscribedTopicDescriberImpl.java

```diff
@@ -79,49 +80,65 @@ public int numPartitions(Uuid topicId) {
      */
     @Override
     public Set<String> racksForPartition(Uuid topicId, int partition) {
+        TopicImage topic = metadataImage.topics().getTopic(topicId);
+        if (topic != null) {
+            PartitionRegistration partitionRegistration = topic.partitions().get(partition);
+            if (partitionRegistration != null) {
+                Set<String> racks = new HashSet<>();
+                for (int replica : partitionRegistration.replicas) {
+                    // Only add the rack if it is available for the broker/replica.
+                    metadataImage.cluster().broker(replica).rack().ifPresent(racks::add);
+                }
+                return racks;
+            }
+        }
         return Set.of();
     }
 
     /**
-     * Returns a set of assignable partitions from the topic metadata.
-     * If the allowed partition map is null, all the partitions in the corresponding
-     * topic metadata are returned for the argument topic id. If allowed map is empty,
+     * Returns a set of assignable partitions from the metadata image.
+     * If the allowed partition map is Optional.empty(), all the partitions in the corresponding
+     * topic image are returned for the argument topic id. If allowed map is empty,
      * empty set is returned.
      *
      * @param topicId The uuid of the topic
      * @return Set of integers if assignable partitions available, empty otherwise.
      */
     @Override
     public Set<Integer> assignablePartitions(Uuid topicId) {
-        TopicMetadata topic = this.topicMetadata.get(topicId);
+        TopicImage topic = metadataImage.topics().getTopic(topicId);
         if (topic == null) {
             return Set.of();
         }
 
-        if (topicPartitionAllowedMap == null) {
-            return IntStream.range(0, topic.numPartitions()).boxed().collect(Collectors.toUnmodifiableSet());
+        if (topicPartitionAllowedMap.isEmpty()) {
+            return Set.copyOf(topic.partitions().keySet());
```

Comment:
Do we really need to make a copy of the Set? The image is immutable so we could just return a non modifiable version. What do you think?

Reply:
IIUC, the production code uses `HashMap` to build partitions in `TopicImage`. From [Map#keySet Java doc](https://devdocs.io/openjdk~17/java.base/java/util/map#keySet()), the change to the set can be reflected in the map. Is it safer to use `Set.copyOf` here?

https://github.com/apache/kafka/blob/f26974b16db7903bd350f06b5796acac52604d4c/metadata/src/main/java/org/apache/kafka/image/TopicDelta.java#L142-L157

Reply:
Yeah, this is correct. However, the TopicImage is never mutated once built. The metadata image is immutable.

Reply:
As @dajac mentioned, maybe we can make `TopicImage` create immutable collection as it is never modified once built.
```java
    public TopicImage(String name,
                      Uuid id,
                      Map<Integer, PartitionRegistration> partitions) {
        this.name = name;
        this.id = id;
        this.partitions = Collections.unmodifiableMap(partitions);
    }
```

with this change, we can return `topic.partitions().keySet()` directly as it is immutable now.

BTW, I have opened https://issues.apache.org/jira/browse/KAFKA-19305 to fix other image classes too

Reply:
Thanks for the review and suggestion. Remove `Set.copyOf` here.